# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [2]:
import pandas as pd;
import sqlite3 as sql;
import warnings;
warnings.simplefilter("ignore")

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [3]:
conn = sql.connect("../Great_Outdoors_Data_SQLite/go_sales.sqlite")

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [4]:
product = pd.read_sql_query("SELECT * FROM product", conn)
product_type = pd.read_sql_query("SELECT * FROM product_type", conn)
product_line = pd.read_sql_query("SELECT * FROM product_line", conn)
sales_staff = pd.read_sql_query("SELECT * FROM sales_staff", conn)
sales_branch = pd.read_sql_query("SELECT * FROM sales_branch", conn)
retailer_site = pd.read_sql_query("SELECT * FROM retailer_site", conn)
country = pd.read_sql_query("SELECT * FROM country", conn)
order_header = pd.read_sql_query("SELECT * FROM order_header", conn)
order_details = pd.read_sql_query("SELECT * FROM order_details", conn)
target = pd.read_sql_query("SELECT * FROM SALES_TARGETData", conn)
returned_item = pd.read_sql_query("SELECT * FROM returned_item", conn)
return_reason = pd.read_sql_query("SELECT * FROM return_reason", conn)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
product["PRODUCTION_COST"] = pd.to_numeric(product["PRODUCTION_COST"])
df = product["PRODUCTION_COST"].apply(lambda x: x < 100 and x > 50)
df = product.loc[df, ["PRODUCT_NAME", "PRODUCTION_COST"]]
df

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
product["MARGIN"] = pd.to_numeric(product["MARGIN"])
df = product["MARGIN"].apply(lambda x: x < 0.2 or x > 0.6)
df = product.loc[df, ["PRODUCT_NAME", "MARGIN"]]
df

PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
df = country["CURRENCY_NAME"].isin(["francs"])
df = country.loc[df, ["COUNTRY"]].sort_values("COUNTRY")
df

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
product["MARGIN"] = pd.to_numeric(product["MARGIN"])
df = product["MARGIN"].apply(lambda x: x > 0.5)
df = product.loc[df, ["INTRODUCTION_DATE"]].drop_duplicates(subset=["INTRODUCTION_DATE"])
df

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [9]:
df = sales_branch[["ADDRESS2", "REGION"]].notnull().all(axis=1)
df = sales_branch.loc[df, ["ADDRESS1", "CITY"]]
df #Foute vraag, 7 rijen niet 28

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

In [10]:
#Dit is geen vraag

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [11]:
df = country["CURRENCY_NAME"].str.contains("dollar")
df = country.loc[df, ["COUNTRY"]].sort_values(by="COUNTRY")
df

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [12]:
df = retailer_site[(retailer_site["POSTAL_ZONE"].str.startswith("D")) & (retailer_site["ADDRESS2"].notnull())].sort_values(by="ADDRESS2")
df.loc[:, ["ADDRESS1", "ADDRESS2", "CITY"]]

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [13]:
df = pd.to_numeric(returned_item["RETURN_QUANTITY"]).sum()
df

14306

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [14]:
df = sales_branch["REGION"].count()
df

15

In [15]:
#Dit is geen vraag

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [16]:
low_margin = product["MARGIN"].min()
high_margin = product["MARGIN"].max()
average_margin: float = round(product["MARGIN"].mean(), 2)
df = pd.DataFrame({"Laagste marge": low_margin, "Hoogste marge": high_margin, "Gemiddelde marge": average_margin}, index=[1])
df

Laagste marge  Hoogste marge  Gemiddelde marge
1           0.17            0.7               0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [17]:
df = retailer_site.isnull().sum()
df["ADDRESS2"]

286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [18]:
order_details["UNIT_COST"] = pd.to_numeric(order_details["UNIT_COST"])
df = order_details.apply(lambda x: x["UNIT_SALE_PRICE"] < x["UNIT_PRICE"], axis=1)
df = order_details.loc[df, ["UNIT_COST"]].mean().round(2)
df

UNIT_COST    121.99
dtype: float64

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [19]:
df = sales_staff.groupby("POSITION_EN", as_index=False).size()
df.index = range(1, len(df) + 1)
df.rename(columns={"size": "Aantal medewerkers per medewerkersfunctie"}, inplace=True)
df

POSITION_EN  Aantal medewerkers per medewerkersfunctie
1                Branch Manager                                         19
2              District Manager                                          4
3               General Manager                                          1
4  Level 1 Sales Representative                                         13
5  Level 2 Sales Representative                                         39
6  Level 3 Sales Representative                                         24
7              Regional Manager                                          2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [20]:
df = sales_staff.groupby("WORK_PHONE", as_index=False).size()
df.rename(columns={"size": "Aantal medewerkers per telefoonnummer"}, inplace=True)
df[df["Aantal medewerkers per telefoonnummer"] > 4]

WORK_PHONE  Aantal medewerkers per telefoonnummer
4     +(44) 121 3505267                                      5
10  +(61) 03 2982  4242                                      5
11  +31 (0)20 692 93 94                                      5
13    +33 1 68 94 52 20                                      5
18     1 (206) 292-0012                                      5
19     1 (305) 557-4810                                      6
20     1 (310) 281-5722                                      5
21     1 (403) 232-5986                                      5
22     1 (416) 493-5595                                      5
23     1 (617) 268-6754                                      6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [21]:
df = retailer_site.merge(country, on="COUNTRY_CODE", how="inner", indicator=True)
data = df["COUNTRY"].isin(["Netherlands"])
df.loc[data, ["ADDRESS1", "CITY"]]

ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [22]:
df = product.merge(product_type, on="PRODUCT_TYPE_CODE", how="inner", indicator=True)
data = df["PRODUCT_TYPE_EN"].isin(["Eyewear"])
df.loc[data, ["PRODUCT_NAME"]]

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [62]:
df = retailer_site.merge(sales_staff, left_on="RETAILER_CODE", how="inner", right_on="SALES_BRANCH_CODE", indicator=True)
subsets: list[str] = ["ADDRESS1", "FIRST_NAME", "LAST_NAME"]
data = df["POSITION_EN"].isin(["Branch Manager"])
df.loc[data, subsets].drop_duplicates(subset=df[subsets]).head(1)

ADDRESS1 FIRST_NAME  LAST_NAME
3  22, rue Saint-Jean      Karin  Bergström

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [63]:
df = sales_staff.merge(order_header, on="SALES_STAFF_CODE", how="left", indicator=True)
data = df["POSITION_EN"].str.contains("Manager")
df.loc[data, ["POSITION_EN", "ORDER_DATE"]].drop_duplicates().fillna("Niks verkocht")

POSITION_EN     ORDER_DATE
421     Branch Manager  Niks verkocht
422    General Manager  Niks verkocht
1276  Regional Manager  Niks verkocht
1643  District Manager  Niks verkocht
5325    Branch Manager     2022-09-09
5326    Branch Manager     2022-10-31
5327    Branch Manager     2022-12-13
5328    Branch Manager     2022-03-13
5329    Branch Manager     2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [64]:
df = product.merge(product_type, on="PRODUCT_TYPE_CODE", how="inner").merge(order_details, on="PRODUCT_NUMBER", how="inner", indicator=True)
data = df["QUANTITY"].apply(lambda x: x >= 750)
df.loc[data, ["PRODUCT_NAME", "PRODUCT_TYPE_EN"]].drop_duplicates(subset=["PRODUCT_NAME"])

PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [65]:
df = product.merge(order_details, on="PRODUCT_NUMBER", how="inner", indicator=True)
df["UNIT_PRICE"] = pd.to_numeric(df["UNIT_PRICE"])
df["UNIT_SALE_PRICE"] = pd.to_numeric(df["UNIT_SALE_PRICE"])
data = df.apply(lambda x: (x["UNIT_PRICE"] - x["UNIT_SALE_PRICE"]) / x["UNIT_PRICE"] > 0.4, axis=1)
df.loc[data, ["PRODUCT_NAME"]].drop_duplicates(subset=["PRODUCT_NAME"])

PRODUCT_NAME
2765   Hailstorm Titanium Woods Set
37446             BugShield Natural
37967               BugShield Spray
38306         BugShield Lotion Lite
38772              BugShield Lotion
39561             BugShield Extreme
40799                Sun Shelter 15
41194                Sun Shelter 30

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [66]:
df = return_reason.merge(returned_item, on="RETURN_REASON_CODE", how="inner").merge(order_details, on="ORDER_DETAIL_CODE", how="inner", indicator=True)
df["RETURN_QUANTITY"] = pd.to_numeric(df["RETURN_QUANTITY"])
data = df.apply(lambda x: (x["RETURN_QUANTITY"] / x["QUANTITY"]) > 0.9, axis=1)
df.loc[data, ["RETURN_DESCRIPTION_EN"]].drop_duplicates(subset=["RETURN_DESCRIPTION_EN"])

RETURN_DESCRIPTION_EN
451   Wrong product ordered
521   Wrong product shipped
583  Unsatisfactory product

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [67]:
df = product.merge(product_type, on="PRODUCT_TYPE_CODE", how="inner", indicator=True).groupby("PRODUCT_TYPE_EN", as_index=False).size()
df.index = range(1, len(df) + 1)
df.rename(columns={"size": "Aantal producten"}, inplace=True)
df

PRODUCT_TYPE_EN  Aantal producten
1             Binoculars                 4
2   Climbing Accessories                 7
3           Cooking Gear                10
4                Eyewear                 5
5              First Aid                 5
6       Golf Accessories                 4
7      Insect Repellents                 5
8                  Irons                 4
9                 Knives                 5
10              Lanterns                12
11            Navigation                 4
12                 Packs                 6
13               Putters                 3
14                  Rope                 4
15                Safety                 4
16         Sleeping Bags                 7
17             Sunscreen                 5
18                 Tents                 6
19                 Tools                 6
20               Watches                 5
21                 Woods                 4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [68]:
df = country.merge(retailer_site, on="COUNTRY_CODE", how="inner", indicator=True).groupby("COUNTRY", as_index=False).size()
df.index = range(1, len(df) + 1)
df.rename(columns={"size": "Aantal vestigingen"}, inplace=True)
df

COUNTRY  Aantal vestigingen
1        Australia                  10
2          Austria                  10
3          Belgium                  10
4           Brazil                   1
5           Canada                  31
6            China                  10
7          Denmark                   6
8          Finland                   5
9           France                  29
10         Germany                  32
11           Italy                  12
12           Japan                  25
13           Korea                  10
14          Mexico                  10
15     Netherlands                  20
16           Spain                   9
17          Sweden                  12
18     Switzerland                  13
19          Taiwan                  10
20  United Kingdom                  25
21   United States                 101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [69]:
df = product.merge(product_type, on="PRODUCT_TYPE_CODE", how="inner").merge(order_details, on="PRODUCT_NUMBER", how="inner", indicator=True)
df["UNIT_PRICE"] = pd.to_numeric(df["UNIT_PRICE"])
data = df["PRODUCT_TYPE_EN"].isin(["Cooking Gear"])
df = df.loc[data].groupby(["PRODUCT_TYPE_EN", "PRODUCT_NAME"], as_index=False).agg({"QUANTITY": "sum", "UNIT_PRICE": "mean"}).round(2).sort_values(by="QUANTITY", ascending=False)
df.index = range(1, len(df) + 1)
df

PRODUCT_TYPE_EN               PRODUCT_NAME  QUANTITY  UNIT_PRICE
1     Cooking Gear        TrailChef Water Bag     41770        6.18
2     Cooking Gear              TrailChef Cup     31834        7.14
3     Cooking Gear          TrailChef Canteen     31712       12.18
4     Cooking Gear      TrailChef Kitchen Kit     25964       22.74
5     Cooking Gear         TrailChef Utensils     15928       17.73
6     Cooking Gear     TrailChef Double Flame     14510      136.60
7     Cooking Gear         TrailChef Cook Set     12700       52.61
8     Cooking Gear     TrailChef Single Flame     10614       65.09
9     Cooking Gear           TrailChef Kettle      9900       12.38
10    Cooking Gear  TrailChef Deluxe Cook Set      3744      123.54

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [70]:
df = sales_branch.merge(sales_staff, on="SALES_BRANCH_CODE", how="inner").merge(target, on="SALES_STAFF_CODE", how="inner", indicator=True)
data = df.apply(lambda x: x["CITY"] == "Amsterdam" and x["SALES_YEAR"] == "2022", axis=1)
df = df.loc[data].groupby(["FIRST_NAME", "LAST_NAME"], as_index=False).agg({"SALES_TARGET": ["min", "max"]})
df.columns = [" ".join(columns) for columns in df.columns.values]
df.rename(columns={"SALES_TARGET min": "Laagste verkooptarget", "SALES_TARGET max": "Hoogste verkooptarget"}, inplace=True)
df.index = range(1, len(df) + 1)
df

FIRST_NAME         LAST_NAME  Laagste verkooptarget Hoogste verkooptarget
1     Belinda  Jansen-Velasquez                  1005                   963
2       Ellen           Shapiro                  1004                   986
3         Jan         Haverkamp                  1048                   993
4       Maria          Laponder                  1000                  9956

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [8]:
df = country.merge(sales_branch, on="COUNTRY_CODE", how="inner", indicator=True)
df = df.groupby(["COUNTRY"], as_index=False).agg({"CITY": ["unique", "size"]})
df.columns = [" ".join(columns) for columns in df.columns.values]
df.rename(columns={"CITY unique": "Verkoper", "CITY size": "Klanten"}, inplace=True)
df = df.explode("Verkoper")
df["Verkoper"].apply(lambda x: ", ".join(map(str, x)))
df.index = range(1, len(df) + 1)
df

COUNTRY           Verkoper  Klanten
1        Australia         Melbourne        1
2          Austria              Wien        1
3          Belgium          Heverlee        1
4           Brazil         São Paulo        1
5           Canada           Calgary        2
6           Canada           Toronto        2
7            China          Shanghai        1
8          Finland            Kuopio        1
9           France              Lyon        2
10          France             Paris        2
11         Germany           Hamburg        2
12         Germany           München        2
13           Italy            Milano        1
14           Japan             Tokyo        2
15           Japan        Osaka City        2
16           Korea             Seoul        1
17          Mexico  Distrito Federal        1
18     Netherlands         Amsterdam        1
19           Spain            Bilbao        1
20          Sweden             Kista        1
21     Switzerland            Zürich        1
22          Taiwan         San Chung        1
23  United Kingdom            London        2
24  United Kingdom        Birmingham        2
25   United States            Boston        4
26   United States           Seattle        4
27   United States       Los Angeles        4
28   United States             Miami        4

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [9]:
df = sales_staff.merge(order_header, on="SALES_STAFF_CODE", how="outer", indicator=True)
df = df[df["ORDER_NUMBER"].isnull()]
df[["FIRST_NAME", "LAST_NAME"]]

FIRST_NAME      LAST_NAME
421      Walter         Taylor
422    Penelope        Tamrine
607       Polly         Linton
881       Gracy        Gellens
1276       Else         Mörike
1277      Frank          Fuchs
1642      Fritz         Hirsch
1643       Jörg          Kunze
1730      Maria      Iacobucci
1975       Kick        Kalkman
2381      Karin      Bergström
2382      Denis           Pagé
2479      Sally          White
2507      Frank        Bretton
2508    Georges  Saint-Germain
2933       Alex      Rodriguez
3351   Samantha          Floyd
3723      Julie          Olsen
3881      Corey         Wright
4347    Estelle        Johnson
4348       Greg        Belding
4484      Julie        Béchard
4659      Lucía          Reyna
4930     Kazumi        Uragome
5103      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [18]:
import numpy as np;

df["MARGIN"] = pd.to_numeric(product["MARGIN"])
data = df["MARGIN"].apply(lambda x: x < np.mean(x))
number_of_products: int = len(data)
average_margin = round(df["MARGIN"].mean(), 2)
df = pd.DataFrame({"Totaal aantal producten": [number_of_products], "Gemiddelde marge samen": [average_margin]}, index=[1])
df #glitched out 59 products and 0.29 average margin

Totaal aantal producten  Gemiddelde marge samen
1                        1                     0.4

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [19]:
df = product.merge(order_details, on="PRODUCT_NUMBER", how="inner").merge(returned_item, on="ORDER_DETAIL_CODE", how="outer", indicator=True)
df["UNIT_PRICE"] = pd.to_numeric(df["UNIT_PRICE"])
data = df["UNIT_PRICE"].apply(lambda x: x > 500)
df.loc[data, ["PRODUCT_NAME"]].drop_duplicates(subset=["PRODUCT_NAME"])

PRODUCT_NAME
33               Hailstorm Steel Irons
36            Hailstorm Titanium Irons
41          Lady Hailstorm Steel Irons
44       Lady Hailstorm Titanium Irons
47        Hailstorm Titanium Woods Set
51           Hailstorm Steel Woods Set
55   Lady Hailstorm Titanium Woods Set
58      Lady Hailstorm Steel Woods Set
132                          Star Dome
136                       Star Gazer 2
138                       Star Gazer 3
141                       Star Gazer 6
263                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [20]:
df = sales_staff
df["Ja/Nee"] = df["POSITION_EN"].str.contains("Manager").replace({True: "Ja", False: "Nee"})
df.index = range(1, len(df) + 1)
df[["LAST_NAME", "Ja/Nee"]]

LAST_NAME Ja/Nee
1    Savolainen    Nee
2           Kim    Nee
3          Choi    Nee
4           Kao    Nee
5          Fang    Nee
..          ...    ...
98     Iglesias    Nee
99        Lopes     Ja
100      Barros    Nee
101   Guimarães    Nee
102    Lindholm    Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [21]:
from datetime import date;
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [22]:
from datetime import datetime;

date_str = "16-8-2013"
date_format = "%d-%m-%Y"
date_obj = datetime.strptime(date_str, date_format)
date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [23]:
df = sales_staff
df.rename(columns={"FIRST_NAME": "Medewerker"}, inplace=True)
df["jaren_in_dienst"] = pd.to_datetime(df["DATE_HIRED"]).apply(lambda x: date.today().year - x.year >= 12).replace({True: "lang in dienst", False: "kort in dienst"})
df[["Medewerker", "jaren_in_dienst"]]

Medewerker jaren_in_dienst
1       Tuomas  kort in dienst
2     Chang-ho  kort in dienst
3      Jung-ho  kort in dienst
4    Kwei-Ping  kort in dienst
5    Chin-Tsai  lang in dienst
..         ...             ...
98       Tomás  kort in dienst
99      Bayard  kort in dienst
100     Viviam  kort in dienst
101    Eduardo  kort in dienst
102     Helena  kort in dienst

[102 rows x 2 columns]